### ISSUE TITLE
ArgRequiredElseHelp is ignored if there are *any* default values

### ISSUE THREAD

**Issue by epage** [epage/clapng#96]
<a href="https://github.com/whitfin"><img src="https://avatars.githubusercontent.com/u/5376378?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [whitfin](https://github.com/whitfin)**
_Tuesday May 01, 2018 at 04:59 GMT_
_Originally opened as https://github.com/clap-rs/clap/issues/1264_

----

### Rust Version

rustc 1.25.0 (84203cac6 2018-03-25)

### Affected Version of clap

2.31.2

### Expected Behavior Summary

If `ArgRequiredElseHelp` is enabled, you should see the help menu if an option is not provided correctly.

### Actual Behavior Summary

If you have two arguments, and one has a default, this the help menu is not correctly printed.

### Steps to Reproduce the issue

Run the code below in a test project via `cargo run`. It won't work (rightly so) because the "second" option is not provided - but you'll get an error instead of the help menu.

Do the same again with the default value removed from "first", and you'll then see a help menu. Regardless of the default value of "first", it should show the help menu both times.

```rust
App::new("")
    // package metadata from cargo
    .name(env!("CARGO_PKG_NAME"))
    .about(env!("CARGO_PKG_DESCRIPTION"))
    .version(env!("CARGO_PKG_VERSION"))

    // arguments and flag details
    .arg(Arg::with_name("second")
        .help("second option")
        .required(true))
    .arg(Arg::with_name("first")
        .help("first option")
        .default_value("something"))

    // settings required for parsing
    .setting(AppSettings::ArgRequiredElseHelp)
```

**Comment by epage** [comment]
<a href="https://github.com/whitfin"><img src="https://avatars.githubusercontent.com/u/5376378?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [whitfin](https://github.com/whitfin)**
_Tuesday May 01, 2018 at 05:04 GMT_

----

I think perhaps https://github.com/kbknapp/clap-rs/blob/master/src/app/validator.rs#L63 should use `!reqs_validated` instead of `matcher.is_empty()`.

I dropped a PR in https://github.com/kbknapp/clap-rs/pull/1265; all tests pass still and the issue appears fixed. I'm not familiar enough to know if this is the best approach, so let me know of any feedback.

Edit: actually that PR appears to render the help menu _too_ frequently now, so closed for the time being :)

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Tuesday Jun 05, 2018 at 01:38 GMT_

----

Sorry for the long wait, I've been away for a few weeks with my day job.

while I agree it's not optimal, a default value *is* a value and therefore clap treats it as such. I'd possibly consider adding some sort of setting such as `UserArgRequiredElseHelp` but I'm not sure it'd quite be worth it as the "fix" for this is to simply not provide a default and use something like, `matches.value_of("my-arg-that-needs-a-default").unwrap_or("default_value")` and just "tell" the user in the help text manually that there is a default value (but to `clap` there isn't).

**Comment by epage** [comment]
<a href="https://github.com/pksunkara"><img src="https://avatars.githubusercontent.com/u/174703?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [pksunkara](https://github.com/pksunkara)**
_Sunday Jul 19, 2020 at 10:37 GMT_

----

@CreepySkeleton you were working on default stuff recently, right? Can you take a look at this last comment?

**Comment by epage** [comment]
<a href="https://github.com/CreepySkeleton"><img src="https://avatars.githubusercontent.com/u/50968528?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [CreepySkeleton](https://github.com/CreepySkeleton)**
_Sunday Jul 19, 2020 at 15:38 GMT_

----

I find myself agreeing with @Kayvlim . Default values aren't ephemeral, they are very real values. Clap treats them in almost exactly the same way as any other value (barring conflict resolution which they are exempt from, or at least should have been).

> Is there any use case where ArgRequiredElseHelp makes sense when an arg with a default_value is present?

That is a good question. My answer would be no, there aren't, and in this case we should panic if `ArgRequiredElseHelp` is set and _any_ of the args has `default_value`.

**Comment by epage** [comment]
<a href="https://github.com/Kayvlim"><img src="https://avatars.githubusercontent.com/u/887736?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Kayvlim](https://github.com/Kayvlim)**
_Wednesday Jul 22, 2020 at 01:06 GMT_

----

@CreepySkeleton I suppose you are only agreeing with how I think clap works, but not with how I think it should be changed?

I don't agree that `ArgRequiredElseHelp` should panic if any arg has a `default_value`. It's a perfectly legitimate use case:

If the user runs my application without arguments, I want them to see the help, but if they run it with the single required argument, I want the application to run with a sensible set of default values for the arguments they *didn't* pass (or "override"). I shouldn't have to `unwrap_or` every non-passed argument to accomplish this when a mechanism for default values is available just because I also want to display the help when no arguments have been passed.

So, I think `ArgRequiredElseHelp` should do exactly the same thing it would if there were no `default_value`s whatsoever. This doesn't make default values ephemeral. Instead, it stops pretending they were passed by the user when they really weren't.

In other words, I don't see default values as values the user positively intended to pass to the application by omission. I see them as values the application needs to assume when the user hasn't said anything about them and `None` is not acceptable.

**Comment by epage** [comment]
<a href="https://github.com/dimo414"><img src="https://avatars.githubusercontent.com/u/1885701?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [dimo414](https://github.com/dimo414)**
_Friday Feb 05, 2021 at 08:05 GMT_

----

I was bitten by this issue myself while trying to add ArgRequiredElseHelp to a CLI. It's wholly unintuitive - both to the user and the developer - that the use of default values would be considered sufficient to bypass this check.
